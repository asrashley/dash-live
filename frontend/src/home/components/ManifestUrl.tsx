import { useComputed } from "@preact/signals";
import { type ReadonlySignal } from "@preact/signals-core";
import { useCallback, useRef } from "preact/hooks";

interface ManifestUrlProps {
   manifestUrl: ReadonlySignal<URL>;
   editable: ReadonlySignal<boolean>;
   setValue: (url: string) => void;
}

export function ManifestUrl({ manifestUrl, editable, setValue }: ManifestUrlProps) {
  const inpElt = useRef<HTMLInputElement>();
  const inputClass = useComputed<string>(() => `form-control${ editable.value ? "" : " d-none"}`);
  const linkClass = useComputed<string>(() => editable.value ? "d-none": "link link-underline-opacity-25");
  const onSubmit = useCallback(
    (ev: Event) => {
      ev.preventDefault();
      setValue(inpElt.current.value);
    },
    [setValue]
  );

  return <div className="manifest-url rounded border">
      <form id="mpd-form" onSubmit={onSubmit}>
        <div className="form-group d-flex flex-row row-field-title">
          <label className="col-form-label fw-semibold me-1" for="id_mpd_url">
            MPD URL :
          </label>
          <div className="flex-fill">
            <input
              type="text"
              value={manifestUrl.value.href}
              name="mpd_url"
              id="id_mpd_url"
              className={inputClass}
              ref={inpElt}
            />
            <a id="dashurl" href={manifestUrl.value.href} className={linkClass}>
             {manifestUrl.value.pathname}{manifestUrl.value.search}
            </a>
          </div>
        </div>
      </form>
  </div>;
}
