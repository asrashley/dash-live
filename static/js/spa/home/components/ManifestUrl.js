import { html } from "htm/preact";

export function ManifestUrl({ manifestUrl }) {
  return html` <div className="manifest-url rounded border">
    <span className="fw-semibold">Manifest URL: </span>
    <a id="dashurl" href="${manifestUrl.value}">
      ${manifestUrl.value.pathname}${manifestUrl.value.search}
    </a>
  </div>`;
}
