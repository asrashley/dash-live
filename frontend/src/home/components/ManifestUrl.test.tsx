import { signal } from "@preact/signals";
import { afterEach, describe, expect, test, vi } from "vitest";
import userEvent from "@testing-library/user-event";

import { renderWithProviders } from "../../test/renderWithProviders";

import { ManifestUrl } from "./ManifestUrl";

describe("ManifestUrl component", () => {
  const manifestUrl = signal<URL>();
  const editable = signal<boolean>(false);
  const setValue = vi.fn();

  afterEach(() => {
    vi.clearAllMocks();
  });

  test("shows manifest link", () => {
    manifestUrl.value = new URL(
      "/dash/live/netflix/manifest_e.mpd",
      document.location.href
    );
    editable.value = false;
    const { asFragment, getBySelector } = renderWithProviders(
      <ManifestUrl
        manifestUrl={manifestUrl}
        editable={editable}
        setValue={setValue}
      />
    );
    const inputElt = getBySelector('#id_mpd_url') as HTMLInputElement;
    const anchorElt = getBySelector('#dashurl');
    expect(inputElt.classList.contains('d-none')).toEqual(true);
    expect(anchorElt.classList.contains('d-none')).toEqual(false);
    expect(asFragment()).toMatchSnapshot();
  });

  test("allows manifest to be edited", async () => {
    const user = userEvent.setup();
    manifestUrl.value = new URL(
      "/dash/live/netflix/manifest_e.mpd",
      document.location.href
    );
    editable.value = true;
    const { asFragment, getBySelector } = renderWithProviders(
      <ManifestUrl
        manifestUrl={manifestUrl}
        editable={editable}
        setValue={setValue}
      />
    );
    const inputElt = getBySelector('#id_mpd_url') as HTMLInputElement;
    const anchorElt = getBySelector('#dashurl');
    expect(inputElt.classList.contains('d-none')).toEqual(false);
    expect(anchorElt.classList.contains('d-none')).toEqual(true);
    expect(inputElt.value).toEqual(manifestUrl.value.href);
    await user.click(inputElt);
    await user.type(inputElt, "?abr=0{enter}");
    expect(inputElt.value).toEqual(`${manifestUrl.value.href}?abr=0`);
    expect(setValue).toHaveBeenCalledTimes(1);
    expect(setValue).toHaveBeenCalledWith(inputElt.value);
    expect(asFragment()).toMatchSnapshot();
  });
});
