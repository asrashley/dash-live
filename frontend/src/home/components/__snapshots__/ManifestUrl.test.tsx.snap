// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ManifestUrl component > allows manifest to be edited 1`] = `
<DocumentFragment>
  <div
    class="manifest-url rounded border"
  >
    <form
      id="mpd-form"
    >
      <div
        class="form-group d-flex flex-row row-field-title"
      >
        <label
          class="col-form-label fw-semibold me-1"
          for="id_mpd_url"
        >
          MPD URL :
        </label>
        <div
          class="flex-fill"
        >
          <input
            class="form-control"
            id="id_mpd_url"
            name="mpd_url"
            type="text"
          />
          <a
            class="d-none"
            href="http://localhost:3000/dash/live/netflix/manifest_e.mpd"
            id="dashurl"
          >
            /dash/live/netflix/manifest_e.mpd
          </a>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;

exports[`ManifestUrl component > shows manifest link 1`] = `
<DocumentFragment>
  <div
    class="manifest-url rounded border"
  >
    <form
      id="mpd-form"
    >
      <div
        class="form-group d-flex flex-row row-field-title"
      >
        <label
          class="col-form-label fw-semibold me-1"
          for="id_mpd_url"
        >
          MPD URL :
        </label>
        <div
          class="flex-fill"
        >
          <input
            class="form-control d-none"
            id="id_mpd_url"
            name="mpd_url"
            type="text"
          />
          <a
            class="link link-underline-opacity-25"
            href="http://localhost:3000/dash/live/netflix/manifest_e.mpd"
            id="dashurl"
          >
            /dash/live/netflix/manifest_e.mpd
          </a>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
